### PR TITLE
feat(warehouse): 497-refact-warehouse-based-on-catalog

### PR DESCRIPTION
**Pull request by MarcusViniciusSS**
* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Other information**:

### CODE REVIEW

File: src/Contracts/Services/Warehouse/Projection.cs [hunk 1]
```diff
@@ -5,7 +5,29 @@ namespace Contracts.Services.Warehouse;
 
 public static class Projection
 {
-    public record Inventory(Guid Id, Guid OwnerId, bool IsDeleted) : IProjection;
+    public record Inventory(Guid Id, Guid OwnerId, bool IsDeleted) : IProjection
```
Comment (AntonioFalcaoJr) [inline src/Contracts/Services/Warehouse/Projection.cs:8]: InventoryGridItem

File: src/Contracts/Services/Warehouse/Projection.cs [hunk 2]
```diff
@@ -5,7 +5,29 @@ namespace Contracts.Services.Warehouse;
...
+    {
+        public static implicit operator Protobuf.InventoryGridItem(Inventory inventory)
+            => new()
+            {
+                InventoryId = inventory.Id.ToString(),
+                OwnerId = inventory.OwnerId.ToString(),
+                IsDeleted = inventory.IsDeleted
```
Comment (AntonioFalcaoJr) [inline src/Contracts/Services/Warehouse/Projection.cs:15]: Remove IsDeleted from proto. It's an internal concern

File: src/Contracts/Services/Warehouse/Projection.cs [hunk 3]
```diff
@@ -5,7 +5,29 @@ namespace Contracts.Services.Warehouse;
...
+            };
+    }
 
-    public record InventoryItem(Guid Id, Guid InventoryId, Dto.Product Product, int Quantity, string Sku, bool IsDeleted) : IProjection;
+    public record InventoryItem(Guid Id, Guid InventoryId, Dto.Product Product, int Quantity, string Sku,
```
Comment (AntonioFalcaoJr) [inline src/Contracts/Services/Warehouse/Projection.cs:19]: InventoryItemListItem

File: src/Contracts/Services/Warehouse/Projection.cs [hunk 4]
```diff
@@ -5,7 +5,29 @@ namespace Contracts.Services.Warehouse;
...
+        bool IsDeleted) : IProjection
+    {
+        public static implicit operator Protobuf.InventoryItemListItem(InventoryItem inventoryItem)
+            => new()
+            {
+                Id = inventoryItem.Id.ToString(),
+                InventoryId = inventoryItem.InventoryId.ToString(),
+                IsDeleted = inventoryItem.IsDeleted,
```
Comment (AntonioFalcaoJr) [inline src/Contracts/Services/Warehouse/Projection.cs:27]: Remove IsDeleted from proto. It's an internal concern

File: src/Contracts/Services/Warehouse/Query.cs [hunk 5]
```diff
@@ -1,10 +1,20 @@
...
-    public record GetInventories(ushort Limit, ushort Offset) : Message, IQuery;
+    public record ListInventoryGridItems(Paging Paging) : IQuery
+    {
+        public static implicit operator ListInventoryGridItems(Protobuf.ListInventoryGridItemsRequest request)
+            => new(request.Paging);
 
-    public record GetInventoryItems(Guid InventoryId, ushort Limit, ushort Offset) : Message(CorrelationId: InventoryId), IQuery;
+    }
+
+    public record ListInventoryItems(Guid InventoryId, Paging Paging) : IQuery
```
Comment (AntonioFalcaoJr) [inline src/Contracts/Services/Warehouse/Query.cs:15]: ListInventoryItemsListItems

File: src/Services/Warehouse/Command/Application/DependencyInjection/Extensions/ServiceCollectionExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    public static IServiceCollection AddCommandInteractors(this IServiceCollection services)
+        => services
+            .AddScoped<IInteractor<Command.CreateInventory>, CreateInventoryInteractor>()
+            .AddScoped<IInteractor<Command.DecreaseInventoryAdjust>, DecreaseInventoryAdjustInteractor>()
+            .AddScoped<IInteractor<Command.IncreaseInventoryAdjust>, IncreaseInventoryAdjustInteractor>()
+            .AddScoped<IInteractor<Command.ReceiveInventoryItem>, ReceiveInventoryItemInteractor>();
+
+    public static IServiceCollection AddEventInteractors(this IServiceCollection services)
+        => services
+            .AddScoped<IInteractor<ShoppingCartDomainEvent.CartItemAdded>, CartItemAddedInteractor>();
```
Comment (AntonioFalcaoJr) [inline src/Services/Warehouse/Command/Application/DependencyInjection/Extensions/ServiceCollectionExtensions.cs:26]: Event Interactor must be injected by the interface, based on broadcast

File: src/Services/Warehouse/Command/Application/DependencyInjection/Extensions/ServiceCollectionExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,27 @@
+using Application.Abstractions;
+using Application.Services;
+using Application.UseCases.Commands;
+using Application.UseCases.Events;
+using Contracts.Services.Warehouse;
+using Microsoft.Extensions.DependencyInjection;
+using ShoppingCartDomainEvent = Contracts.Services.ShoppingCart.DomainEvent;
```
Comment (AntonioFalcaoJr) [inline src/Services/Warehouse/Command/Application/DependencyInjection/Extensions/ServiceCollectionExtensions.cs:7]:
```suggestion
using ShoppingCart = Contracts.Services.ShoppingCart;
```

File: src/Services/Warehouse/Command/Application/DependencyInjection/Extensions/ServiceCollectionExtensions.cs (hunk 6 above)
Comment (AntonioFalcaoJr) [inline src/Services/Warehouse/Command/Application/DependencyInjection/Extensions/ServiceCollectionExtensions.cs:26]:
```suggestion
            .AddScoped<IInteractor<ShoppingCart.DomainEvent.CartItemAdded>, CartItemAddedInteractor>();
```

File: src/Services/Warehouse/Command/Infrastructure.MessageBus/PipeFilters/ContractValidatorFilter.cs [hunk 8]
```diff
@@ -34,7 +34,7 @@ public async Task Send(ConsumeContext<T> context, IPipe<ConsumeContext<T>> next)
         Log.Error("Contract validation errors: {Errors}", validationResult.Errors);
 
         await context.Send(
-            destinationAddress: new($"queue:warehouse.{KebabCaseEndpointNameFormatter.Instance.SanitizeName(typeof(T).Name)}.contract-errors"),
+            destinationAddress: new($"queue:catalog.{KebabCaseEndpointNameFormatter.Instance.SanitizeName(typeof(T).Name)}.contract-errors"),
```
Comment (AntonioFalcaoJr) [inline src/Services/Warehouse/Command/Infrastructure.MessageBus/PipeFilters/ContractValidatorFilter.cs:37]:
```suggestion
            destinationAddress: new($"queue:warehouse.{KebabCaseEndpointNameFormatter.Instance.SanitizeName(typeof(T).Name)}.contract-errors"),
```

File: src/Services/Warehouse/Query/Application/DependencyInjection/ServiceCollectionExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Application.DependencyInjection;
+
+public static class ServiceCollectionExtensions
+{
+    public static IServiceCollection AddEventInteractors(this IServiceCollection services)
+        => services
+            .AddScoped<IProjectInventoryDetailsWhenChangedInteractor, ProjectInventoryDetailsWhenChangedInteractor>()
+            .AddScoped<IProjectInventoryItemWhenChangedInteractor, ProjectInventoryItemWhenChangedInteractor>();
```
Comment (AntonioFalcaoJr) [inline src/Services/Warehouse/Query/Application/DependencyInjection/ServiceCollectionExtensions.cs:15]: Interactor names must match with the projections and protobuf

File: src/Services/Warehouse/Query/Application/DependencyInjection/ServiceCollectionExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+    public static IServiceCollection AddQueryInteractors(this IServiceCollection services)
+        => services
+            .AddScoped<IInteractor<Query.ListInventoryGridItems, IPagedResult<Projection.Inventory>>, ListInventoriesGridInteractor>()
+            .AddScoped<IInteractor<Query.ListInventoryItems, IPagedResult<Projection.InventoryItem>>, ListInventoryItemsInteractor>();
```
Comment (AntonioFalcaoJr) [inline src/Services/Warehouse/Query/Application/DependencyInjection/ServiceCollectionExtensions.cs:20]: Query names must match with the projections and protobuf

File: src/Services/Warehouse/Query/Infrastructure.EventBus/DependencyInjection/Providers/ILazyInteractorProvider.cs [hunk 11]
```diff
@@ -0,0 +1,8 @@
+﻿using Contracts.Abstractions.Messages;
+
+namespace Infrastructure.EventBus.DependencyInjection.Providers;
+
+public interface ILazyInteractorProvider
+{
+    Task InteractAsync<T>(T message, CancellationToken cancellationToken) where T : IEvent;
+}
```
Comment (AntonioFalcaoJr) [inline src/Services/Warehouse/Query/Infrastructure.EventBus/DependencyInjection/Providers/ILazyInteractorProvider.cs:8]: remove

File: src/Services/Warehouse/Query/Infrastructure.EventBus/DependencyInjection/Providers/LazyInteractorProvider.cs [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+    public LazyInteractorProvider(IServiceProvider serviceProvider)
+    {
+        _serviceProvider = serviceProvider;
+    }
+
+    public async Task InteractAsync<T>(T message, CancellationToken cancellationToken) where T : IEvent 
+        => await _serviceProvider
+            .GetRequiredService<IInteractor<T>>()
+            .InteractAsync(message, cancellationToken);
+}
```
Comment (AntonioFalcaoJr) [inline src/Services/Warehouse/Query/Infrastructure.EventBus/DependencyInjection/Providers/LazyInteractorProvider.cs:20]: remove

File: src/Services/Warehouse/Query/Infrastructure.EventBus/PipeFilters/ContractValidatorFilter.cs [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+        if (validationResult.IsValid)
+        {
+            await next.Send(context);
+            return;
+        }
+
+        Log.Error("Contract validation errors: {Errors}", validationResult.Errors);
+
+        await context.Send(
+            destinationAddress: new($"queue:catalog.{KebabCaseEndpointNameFormatter.Instance.SanitizeName(typeof(T).Name)}.contract-errors"),
```
Comment (AntonioFalcaoJr) [inline src/Services/Warehouse/Query/Infrastructure.EventBus/PipeFilters/ContractValidatorFilter.cs:37]: warehouse

File: src/Web/WebAPI/APIs/Warehouses/WarehouseApi.cs [hunk 14]
```diff
@@ -13,14 +14,15 @@ public static IVersionedEndpointRouteBuilder MapWarehouseApiV1(this IVersionedEn
...
+            => ApplicationApi.ListAsync<WarehouseService.WarehouseServiceClient, InventoryGridItem>
+                (request, (client, ct) => client.ListInventoryGridItemsAsync(new ListInventoryGridItemsRequest(), cancellationToken: ct)));
 
         group.MapPost("/", ([AsParameters] Requests.CreateInventory request)
             => ApplicationApi.SendCommandAsync<Command.CreateInventory>(request));
 
-        group.MapGet("/{inventoryId:guid}/items", (IBus bus, Guid inventoryId, ushort? limit, ushort? offset, CancellationToken cancellationToken)
-            => ApplicationApi.GetPagedProjectionAsync<Query.GetInventoryItems, Projection.InventoryItem>(bus, new(inventoryId, limit ?? 0, offset ?? 0), cancellationToken));
+        // group.MapGet("/{inventoryId:guid}/items", (IBus bus, Guid inventoryId, ushort? limit, ushort? offset, CancellationToken cancellationToken)
+        //     => ApplicationApi.GetPagedProjectionAsync<Query.ListInventoryItems, Projection.InventoryItem>(bus, new(inventoryId, limit ?? 0, offset ?? 0), cancellationToken));
```
Comment (AntonioFalcaoJr) [inline src/Web/WebAPI/APIs/Warehouses/WarehouseApi.cs:25]: ???

File: src/Web/WebAPI/APIs/Warehouses/WarehouseApi.cs [hunk 15]
```diff
@@ -13,14 +14,15 @@ public static IVersionedEndpointRouteBuilder MapWarehouseApiV1(this IVersionedEn
     {
         var group = builder.MapGroup(BaseUrl).HasApiVersion(1);
 
-        group.MapGet("/", (IBus bus, ushort limit, ushort offset, CancellationToken cancellationToken)
-            => ApplicationApi.GetPagedProjectionAsync<Query.GetInventories, Projection.Inventory>(bus, new(limit, offset), cancellationToken));
+        group.MapGet("/", ([AsParameters] Requests.ListInventoryGridItems request)
+            => ApplicationApi.ListAsync<WarehouseService.WarehouseServiceClient, InventoryGridItem>
+                (request, (client, ct) => client.ListInventoryGridItemsAsync(new ListInventoryGridItemsRequest(), cancellationToken: ct)));
```
Comment (AntonioFalcaoJr) [inline src/Web/WebAPI/APIs/Warehouses/WarehouseApi.cs:19]: request here

File: src/Web/WebAPI/APIs/Warehouses/WarehouseApi.cs [hunk 16]
```diff
@@ -38,8 +40,9 @@ public static IVersionedEndpointRouteBuilder MapWarehouseApiV2(this IVersionedEn
     {
         var group = builder.MapGroup(BaseUrl).HasApiVersion(2);
 
-        group.MapGet("/", (IBus bus, ushort limit, ushort offset, CancellationToken cancellationToken)
-            => ApplicationApi.GetPagedProjectionAsync<Query.GetInventories, Projection.Inventory>(bus, new(limit, offset), cancellationToken));
+        group.MapGet("/", ([AsParameters] Requests.ListInventoryGridItems request)
+            => ApplicationApi.ListAsync<WarehouseService.WarehouseServiceClient, InventoryGridItem>
+                (request, (client, ct) => client.ListInventoryGridItemsAsync(new ListInventoryGridItemsRequest(), cancellationToken: ct)));
```
Comment (AntonioFalcaoJr) [inline src/Web/WebAPI/APIs/Warehouses/WarehouseApi.cs:45]: request instance

**Review by AntonioFalcaoJr — CHANGES_REQUESTED** [review]

File: src/Contracts/Services/Warehouse/Projection.cs (hunk 1 above)
Comment (MarcusViniciusSS) [reply src/Contracts/Services/Warehouse/Projection.cs:8]: ok

File: src/Services/Warehouse/Query/Infrastructure.EventBus/DependencyInjection/Providers/LazyInteractorProvider.cs (hunk 12 above)
Comment (MarcusViniciusSS) [reply src/Services/Warehouse/Query/Infrastructure.EventBus/DependencyInjection/Providers/LazyInteractorProvider.cs]: removir

File: src/Contracts/Services/Warehouse/Projection.cs (hunk 3 above)
Comment (MarcusViniciusSS) [reply src/Contracts/Services/Warehouse/Projection.cs:19]: ok

File: src/Contracts/Services/Warehouse/Projection.cs (hunk 4 above)
Comment (MarcusViniciusSS) [reply src/Contracts/Services/Warehouse/Projection.cs:27]: ok

File: src/Services/Warehouse/Query/Infrastructure.EventBus/PipeFilters/ContractValidatorFilter.cs (hunk 13 above)
Comment (MarcusViniciusSS) [reply src/Services/Warehouse/Query/Infrastructure.EventBus/PipeFilters/ContractValidatorFilter.cs:37]: done

File: src/Web/WebAPI/APIs/Warehouses/WarehouseApi.cs (hunk 14 above)
Comment (MarcusViniciusSS) [reply src/Web/WebAPI/APIs/Warehouses/WarehouseApi.cs:25]: Yet not been implemented.

File: src/Web/WebAPI/APIs/Warehouses/WarehouseApi.cs (hunk 15 above)
Comment (MarcusViniciusSS) [reply src/Web/WebAPI/APIs/Warehouses/WarehouseApi.cs]: ok, fix it.

File: src/Services/Warehouse/Query/Application/DependencyInjection/ServiceCollectionExtensions.cs (hunk 9 above)
Comment (MarcusViniciusSS) [reply src/Services/Warehouse/Query/Application/DependencyInjection/ServiceCollectionExtensions.cs:15]: ok, I'll refactor it.

File: src/Services/Warehouse/Command/Application/DependencyInjection/Extensions/ServiceCollectionExtensions.cs (hunk 6 above)
Comment (MarcusViniciusSS) [reply src/Services/Warehouse/Command/Application/DependencyInjection/Extensions/ServiceCollectionExtensions.cs:26]: done

File: src/Services/Warehouse/Query/Infrastructure.EventBus/DependencyInjection/Providers/ILazyInteractorProvider.cs (hunk 11 above)
Comment (MarcusViniciusSS) [reply src/Services/Warehouse/Query/Infrastructure.EventBus/DependencyInjection/Providers/ILazyInteractorProvider.cs]: done

File: src/Contracts/Services/Warehouse/Query.cs [hunk 17]
```diff
@@ -1,10 +1,20 @@
...
+    {
+        public static implicit operator ListInventoryGridItems(Protobuf.ListInventoryGridItemsRequest request)
+            => new(request.Paging);
 
-    public record GetInventoryItems(Guid InventoryId, ushort Limit, ushort Offset) : Message(CorrelationId: InventoryId), IQuery;
+    }
+
+    public record ListInventoryItemsListItems(Guid InventoryId, Paging Paging) : IQuery
+    {
+        public static implicit operator ListInventoryItemsListItems(Protobuf.ListInventoryItemsRequest request)
```
Comment (AntonioFalcaoJr) [inline src/Contracts/Services/Warehouse/Query.cs:17]: Protobuf.ListInventoryItemsListItemsResquest

File: src/Services/Warehouse/Command/Application/DependencyInjection/Extensions/ServiceCollectionExtensions.cs [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+    public static IServiceCollection AddCommandInteractors(this IServiceCollection services)
+        => services
+            .AddScoped<IInteractor<Command.CreateInventory>, CreateInventoryInteractor>()
+            .AddScoped<IInteractor<Command.DecreaseInventoryAdjust>, DecreaseInventoryAdjustInteractor>()
+            .AddScoped<IInteractor<Command.IncreaseInventoryAdjust>, IncreaseInventoryAdjustInteractor>()
+            .AddScoped<IInteractor<Command.ReceiveInventoryItem>, ReceiveInventoryItemInteractor>();
+
+    public static IServiceCollection AddEventInteractors(this IServiceCollection services)
+        => services
+            .AddScoped<ICartItemAddedInteractor, CartItemAddedInteractor>();
```
Comment (AntonioFalcaoJr) [inline src/Services/Warehouse/Command/Application/DependencyInjection/Extensions/ServiceCollectionExtensions.cs:26]: missing name pattern

File: src/Services/Warehouse/Query/Infrastructure.EventBus/Consumers/Events/ProjectInventoryItemWhenChangedConsumer.cs [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace Infrastructure.EventBus.Consumers.Events;
+
+public class ProjectInventoryItemWhenChangedConsumer :
+    IConsumer<DomainEvent.InventoryAdjustmentDecreased>,
+    IConsumer<DomainEvent.InventoryAdjustmentIncreased>,
+    IConsumer<DomainEvent.InventoryItemIncreased>,
+    IConsumer<DomainEvent.InventoryItemReceived>
+{
+    private readonly IProjectInventoryItemListItemWhenInventoryItemChangedInteractor _interactor;
```
Comment (AntonioFalcaoJr) [inline src/Services/Warehouse/Query/Infrastructure.EventBus/Consumers/Events/ProjectInventoryItemWhenChangedConsumer.cs:13]:
```suggestion
    private readonly IProjectInventoryItemListItemWhenInventoryChangedInteractor _interactor;
```

File: src/Services/Warehouse/Query/Infrastructure.EventBus/Consumers/Events/ProjectInventoryItemWhenChangedConsumer.cs [hunk 20]
```diff
@@ -0,0 +1,31 @@
+using Application.UseCases.Events;
+using Contracts.Services.Warehouse;
+using MassTransit;
+
+namespace Infrastructure.EventBus.Consumers.Events;
+
+public class ProjectInventoryItemWhenChangedConsumer :
```
Comment (AntonioFalcaoJr) [inline src/Services/Warehouse/Query/Infrastructure.EventBus/Consumers/Events/ProjectInventoryItemWhenChangedConsumer.cs:7]:
```suggestion
public class ProjectInventoryItemListItemWhenInventoryChangedConsumer  :
```

File: src/Web/WebAPI/APIs/Warehouses/Requests.cs [hunk 21]
```diff
@@ -36,4 +37,18 @@ public record DecreaseInventoryAdjust(IBus Bus, Guid InventoryId, Guid ItemId, i
...
     }
+    
+    public record ListInventoryGridItems(WarehouseService.WarehouseServiceClient Client, int? Limit, int? Offset, CancellationToken CancellationToken)
+        : Validatable<ListInventoryGridItemsValidator>, IQueryRequest<WarehouseService.WarehouseServiceClient>
+    {
+        public static implicit operator ListInventoryGridItemsRequest(ListInventoryGridItems request)
+            => new() { Paging = new() { Limit = request.Limit, Offset = request.Offset } };
+    }
+    
+    public record  ListInventoryItems(WarehouseService.WarehouseServiceClient Client, Guid InventoryId, int? Limit, int? Offset, CancellationToken CancellationToken)
```
Comment (AntonioFalcaoJr) [inline src/Web/WebAPI/APIs/Warehouses/Requests.cs:48]:
```suggestion
    public record  ListInventoryGridItems(WarehouseService.WarehouseServiceClient Client, Guid InventoryId, int? Limit, int? Offset, CancellationToken CancellationToken)
```

**Review by AntonioFalcaoJr — CHANGES_REQUESTED** [review]

File: src/Services/Warehouse/Command/Application/DependencyInjection/Extensions/ServiceCollectionExtensions.cs (hunk 18 above)
Comment (MarcusViniciusSS) [reply src/Services/Warehouse/Command/Application/DependencyInjection/Extensions/ServiceCollectionExtensions.cs]: fixed

**Review by AntonioFalcaoJr — APPROVED** [review]

**Comment by frankyjquintero** [comment]
Docker Services error: New path

https://github.com/AntonioFalcaoJr/EDA.CleanArch.DDD.CQRS.EventSourcing/blob/ccc84bc90c0cffb8dd580fbc9cbf3567915a50ce/docker-compose.Staging.Services.yaml#L289

**Pull request merged**